### PR TITLE
Make navigator.connection attribute available via a feature.

### PR DESCRIPTION
**Pull request by mkarolin**
Fixes brave/brave-browser#20122

## Submitter Checklist:

- [ ] I confirm that no security/privacy review [is needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or that I have [requested](https://github.com/brave/security/issues/new/choose) one
- [ ] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [ ] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [ ] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [ ] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [ ] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [ ] Checked the PR locally: `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests`, `npm run lint`, `npm run gn_check`, `npm run tslint`
- [ ] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

### CODE REVIEW

File: chromium_src/third_party/blink/renderer/DEPS [hunk 1]
```diff
@@ -1,4 +1,5 @@
 include_rules = [
+  "+../gen/third_party/blink/renderer/bindings/modules/v8",
```
Comment (goodov) [inline chromium_src/third_party/blink/renderer/DEPS:2]: I think we should automate `+..gen/` DEPS rule as we do for `src/` now.
Please add this:
```
      # Add `+../gen/...` rule.
      root_src_relative_dir = '../gen/' + posixpath.relpath(
          root_src_dir_path_norm, NormalizePath(self.base_directory))
      directory_rules.AddRule('+' + root_src_relative_dir,
                              root_src_relative_dir,
                              'Gen rule for ' + root_src_relative_dir)
```
[right after this line](https://github.com/brave/brave-core/blob/587a51ad4f2508e650d3ddef8008404da3c6512b/chromium_src/python_modules/buildtools/checkdeps/brave_builddeps.py#L32).

And cleanup these two files manually:
`brave/chromium_src/chrome/grit/DEPS`
`brave/chromium_src/third_party/blink/renderer/DEPS`

File: chromium_src/third_party/blink/renderer/platform/bindings/idl_member_installer.cc [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+#include "src/third_party/blink/renderer/platform/bindings/idl_member_installer.cc"
+
+#include "base/strings/string_piece.h"
+#include "third_party/blink/public/common/features.h"
+
+namespace blink {
+
+namespace bindings {
+
+constexpr char kConnection[] = "connection";
```
[the remark this answers is not in the document]
Comment (mkarolin) [reply chromium_src/third_party/blink/renderer/platform/bindings/idl_member_installer.cc]: Fixed in 2e9541c

File: chromium_src/third_party/blink/renderer/platform/bindings/idl_member_installer.cc [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+void IDLMemberInstaller::BraveInstallAttributes(
+    v8::Isolate* isolate,
+    const DOMWrapperWorld& world,
+    v8::Local<v8::Template> instance_template,
+    v8::Local<v8::Template> prototype_template,
+    v8::Local<v8::Template> interface_template,
+    v8::Local<v8::Signature> signature,
+    base::span<const AttributeConfig> configs) {
+  for (const auto& config : configs) {
+    if (!base::FeatureList::IsEnabled(
```
[the remark this answers is not in the document]
Comment (mkarolin) [reply chromium_src/third_party/blink/renderer/platform/bindings/idl_member_installer.cc]: Fixed in 2e9541c

File: chromium_src/third_party/blink/renderer/platform/bindings/idl_member_installer.cc [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+            blink::features::kNavigatorConnectionAttribute) &&
+        base::StringPiece(config.name).compare(kConnection) == 0) {
```
[the remark this answers is not in the document]
Comment (mkarolin) [reply chromium_src/third_party/blink/renderer/platform/bindings/idl_member_installer.cc]: Fixed in 2e9541c

File: renderer/test/navigator_connection_attribute_browsertest.cc [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+                       IsAttributeAvailable) {
+  EXPECT_EQ(IsNavigatorConnectionAttributeEnabled(),
+            base::FeatureList::IsEnabled(
+                blink::features::kNavigatorConnectionAttribute));
+
+  const GURL url = https_server_.GetURL("/simple.html");
+  ASSERT_TRUE(ui_test_utils::NavigateToURL(browser(), url));
+
+  auto result = content::EvalJs(main_frame(), "'connection' in navigator");
+  EXPECT_EQ(result.value.GetBool(), IsNavigatorConnectionAttributeEnabled());
```
[the remark this answers is not in the document]
Comment (mkarolin) [reply renderer/test/navigator_connection_attribute_browsertest.cc]: Fixed in 2e9541c

File: chromium_src/third_party/blink/renderer/platform/bindings/idl_member_installer.cc [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+      continue;
```
[the remark this answers is not in the document]
Comment (mkarolin) [reply chromium_src/third_party/blink/renderer/platform/bindings/idl_member_installer.cc:37]: I don't see anything passed into here that could be used to identify the caller as the `navigator`. Instead, in 2e9541c, I made the methods private and added `blink::V8Navigator` as a friend. Also, added a comment to use caution when adding another filter.

[the remark this answers is not in the document]
Comment (goodov) [reply chromium_src/third_party/blink/renderer/platform/bindings/idl_member_installer.cc:37]: how about a version with templates? we don't need to change a header every time, just need to add new handlers in `.cc` and we're good.
https://gist.github.com/goodov/81da0d6a8a779f41d1dc17bccb4d4dd9
tested locally, works fine.

[the remark this answers is not in the document]
Comment (mkarolin) [reply chromium_src/third_party/blink/renderer/platform/bindings/idl_member_installer.cc:37]: Great idea, thanks 👍 Will update.

[the remark this answers is not in the document]
Comment (mkarolin) [reply chromium_src/third_party/blink/renderer/platform/bindings/idl_member_installer.cc:32]: Applied your patch and squashed fixes into 28ef508298a07c5fa6d73e688d2d8afc386adcc9

**Review by goodov — APPROVED** [review]

**Pull request merged**